frontend/src/routes/(authenticated)/project/[project_code]/BulkAddProjectMembers.svelte

```diff
@@ -0,0 +1,110 @@
+<script lang="ts">
+  import { BadgeButton, MemberBadge } from '$lib/components/Badges';
+  import { DialogResponse, FormModal } from '$lib/components/modals';
+  import { Input, ProjectRoleSelect, TextArea, passwordFormRules } from '$lib/forms';
+  import { ChangeProjectDescriptionDocument, ProjectRole } from '$lib/gql/types';
+  import t from '$lib/i18n';
+  import { z } from 'zod';
+  import { _bulkAddProjectMembers } from './+page';
+  import { hash } from '$lib/util/hash';
+  import { AdminContent } from '$lib/layout';
+  import { createEventDispatcher } from 'svelte';
+  import Icon from '$lib/icons/Icon.svelte';
+
+  enum BulkAddSteps {
+    Add,
+    Results,
+  }
+
+  let currentStep = BulkAddSteps.Add;
+
+  const dispatch = createEventDispatcher();
+
+  export let projectId: string;
+  const schema = z.object({
+    usernamesText: z.string().min(1, $t('register.name_missing')),
+    password: passwordFormRules($t),
+  });
+
+  let formModal: FormModal<typeof schema>;
+  $: form = formModal?.form();
+
+  let createdCount: number | undefined = undefined;
+  let usernameConflicts: string[] = [];
+
+  const usernameRe = /^[a-zA-Z0-9_]+$/;
+
+  function validateUsernames(usernames: string[]): boolean {
+    return usernames.every(s => usernameRe.test(s));
+  }
+
+  async function openModal(): Promise<void> {
+    const { response, formState } = await formModal.open(async () => {
+      const passwordHash = await hash($form.password);
+      const usernames = $form.usernamesText.split('\n').map(s => s.trim());
+      if (!validateUsernames(usernames)) {
+        return $t('project_page.bulk_add_members.usernames_alphanum_only');
+      }
+      const { error, data } = await _bulkAddProjectMembers({
+        projectId,
+        passwordHash,
+        usernames,
+        role: ProjectRole.Editor, // Managers not allowed to have shared passwords
+      });
+
+      createdCount = data?.bulkAddProjectMembers.bulkAddProjectMembersResult?.createdCount;
+      usernameConflicts = data?.bulkAddProjectMembers.bulkAddProjectMembersResult?.usernameConflicts ?? [];
+      return error?.message;
+    });
+    if (response === DialogResponse.Submit) {
+      if (currentStep < BulkAddSteps.Results) {
+        // Go to next page
+        currentStep++;
+        await openModal();
+      } else {
+        currentStep = BulkAddSteps.Add;
+        dispatch('bulkCreated', createdCount);
+        return;
+      }
+    }
+  }
+</script>
+
+<AdminContent>
+  <BadgeButton type="badge-success" icon="i-mdi-account-plus-outline" on:click={openModal}>
+    {$t('project_page.bulk_add_members.add_button')}
+  </BadgeButton>
+
+  <FormModal bind:this={formModal} {schema} let:errors>
+    <span slot="title">{$t('project_page.bulk_add_members.modal_title')}</span>
```

Comment:
So far we've always had a label directly under the title, which had some top-padding. We don't in this case, which finally makes it visible, that the title should probably have a bit of a bottom margin. `bm-4` maybe? I'd put it deep in the modal component unless it looks bad on our existing modals.

Reply:
Done in commit e1391d6498280e77fd10d35730508154d351e23b, in the FormModal component since that's where the title span slot is handled. I've checked other modals and it seems to look fine.